pkg/datalayer/datalayer.go

```diff
@@ -1,28 +1,172 @@
 package datalayer
 
 import (
+	"encoding/json"
+	"fmt"
 	"os"
+	"os/exec"
 	"path/filepath"
+	"strings"
+
+	"github.com/nu7hatch/gouuid"
 )
 
 // ClusterHQ data layer, naive vfs (directory-based) implementation
 
 type DataLayer struct {
-	BasePath string
+	basePath string
+}
+
+type CommitId string
+type CommitMessage string
+
+type Commit struct {
+	Id      CommitId      `json:"id"`
+	Message CommitMessage `json:"message"`
+}
+
+func NewDataLayer(basePath string) *DataLayer {
+	return &DataLayer{basePath}
+}
+
+func (dl *DataLayer) volumePath(volumeName string) string {
+	return filepath.FromSlash(dl.basePath + "/" + volumeName)
+}
+
+func (dl *DataLayer) variantPath(volumeName, variantName string) string {
+	return filepath.FromSlash(dl.basePath + "/" + volumeName + "/branches/" + variantName)
+}
+
+func (dl *DataLayer) commitPath(volumeName string, commitId CommitId) string {
+	return filepath.FromSlash(dl.basePath + "/" + volumeName + "/commits/" + string(commitId))
 }
 
 func (dl *DataLayer) CreateVolume(volumeName string) error {
-	volumePath := filepath.FromSlash(dl.BasePath + "/" + volumeName)
+	volumePath := dl.volumePath(volumeName)
 	return os.MkdirAll(volumePath, 0777)
 }
 
 func (dl *DataLayer) RemoveVolume(volumeName string) error {
-	volumePath := filepath.FromSlash(dl.BasePath + "/" + volumeName)
+	volumePath := filepath.FromSlash(dl.basePath + "/" + volumeName)
 	return os.RemoveAll(volumePath)
 }
 
 func (dl *DataLayer) CreateVariant(volumeName, variantName string) error {
 	// XXX Variants are meant to be tagged commits???
-	variantPath := filepath.FromSlash(dl.BasePath + "/" + volumeName + "/branches/" + variantName)
+	variantPath := filepath.FromSlash(dl.basePath + "/" + volumeName + "/branches/" + variantName)
 	return os.MkdirAll(variantPath, 0777)
 }
+
+func (dl *DataLayer) sanitizePath(path string) error {
+	// Calculate that dl.basePath is a strict prefix of filepath.Clean(path)
+	if !strings.HasPrefix(filepath.Clean(path), dl.basePath) {
```

Comment:
This is vulnerable to the passed `dl.basePath` not being clean. I will leave a comment on that around its assignment.